functorch/experimental/output_dtype.py

```diff
@@ -0,0 +1,136 @@
+from torch._functorch.eager_transforms import (
+    _unwrap_all_tensors_from_functional,
+    _wrap_all_tensors_to_functional,
+)
+from torch._ops import HigherOrderOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+
+
+output_dtype = HigherOrderOperator("output_dtype")
+output_dtype.fallthrough(DispatchKey.PythonDispatcher)
+output_dtype.fallthrough(DispatchKey.PythonTLSSnapshot)
+output_dtype.fallthrough(DispatchKey.ADInplaceOrView)
+output_dtype.fallthrough(DispatchKey.BackendSelect)
+output_dtype.fallthrough(DispatchKey.AutocastCPU)
+
+
+def trace_output_dtype(proxy_mode, func_overload, op, out_dtype, *args):
+    if not isinstance(op, torch._ops.OpOverload):
+        raise ValueError("output_dtype's first argument must be an OpOverload")
+
+    with disable_proxy_modes_tracing():
+        casted_args = pytree.tree_map_only(
+            torch.Tensor, lambda arg: arg.to(dtype=out_dtype), args
+        )
+        out = op(*casted_args)
+
+    node_args = (op, out_dtype, *args)
+    proxy_args = pytree.tree_map(partial(unwrap_proxy, proxy_mode), node_args)
+    out_proxy = proxy_mode.tracer.create_proxy(
+        "call_function", func_overload, proxy_args, {}, name="output_dtype"
+    )
+    return track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
+
+
+@output_dtype.py_impl(DispatchKey.CompositeExplicitAutograd)
+def output_dtype_dense(
+    op: torch._ops.OpOverload,
+    out_dtype: torch.dtype,
+    *args
+):
+    casted_args = pytree.tree_map_only(
+        torch.Tensor, lambda arg: arg.to(dtype=out_dtype), args
+    )
+    res = op(*casted_args)
+    return res
+
+
+@output_dtype.py_impl(DispatchKey.Autograd)
+def output_dtype_autograd(
+    op: torch._ops.OpOverload,
+    out_dtype: torch.dtype,
+    *args
+):
+    # TODO: support autograd
+    flat_operands, _ = pytree.tree_flatten(args)
+    assert all(
+        [not f.requires_grad for f in flat_operands if isinstance(f, torch.Tensor)]
+    )
+
+    _ = torch._C.ExcludeDispatchKeyGuard(
+        torch._C.DispatchKeySet(torch._C.DispatchKey.AutogradCPU)
+    )
+    return output_dtype(op, out_dtype, *args)
+
+
+@output_dtype.py_impl(ProxyTorchDispatchMode)
+def output_dtype_proxy(
+    op: torch._ops.OpOverload,
+    out_dtype: torch.dtype,
+    *args
+):
+    mode = _get_current_dispatch_mode()
+    assert (mode is not None), "Mode should always be enabled for python fallback key"
+    with _pop_mode_temporarily() as mode:
+        if mode.enable_tracing:
+            return trace_output_dtype(mode, output_dtype, op, out_dtype, *args)
+        else:
+            return output_dtype(op, out_dtype, *args)
+
+
+@output_dtype.py_impl(FakeTensorMode)
+def output_dtype_fake_tensor_mode(
+    op: torch._ops.OpOverload,
+    out_dtype: torch.dtype,
+    *args
+):
+    return output_dtype_dense(op, out_dtype, *args)
+
+
+@output_dtype.py_impl(torch._C.DispatchKey.Functionalize)
+def output_dtype_func(op, out_dtype, *args):
```

Comment:
- For safety, we should probably assert in all implementations that op is a torch._ops.OpOverload. For regular Dispatcher operators, we do a parsing step when invoking the operator, but there's not really a good way to do that with HigherOrderOperator right now
- As written the functionalization rule is incorrect. Let's say `op` is `torch.sin_`: then the functionalization rule should invoke `output_dtype(torch.sin, out_dtype, *args)` at some point. Do you see situations where we will need `output_dtype` for in-place or view operations? For now, I would recommend just asserting that `op` has a functional schema, and not worry about the functionalization rule.